metals/src/main/scala/scala/meta/internal/metals/debug/DebugProvider.scala

```diff
@@ -924,4 +920,84 @@ object DebugProvider {
       extends Exception(
         "Build misconfiguration. No semanticdb can be found for you file, please check the doctor."
       )
+
+  sealed abstract class ClassSearch[A](debugProvider: DebugProvider)(implicit
+      ec: ExecutionContext
+  ) {
+    protected def f(): Try[A]
+    protected def dapSessionParams(res: A): Future[DebugSessionParams]
+    private def maybeCreateDapSession(
+        res: Future[Option[A]]
+    ): Future[Option[DebugSession]] = {
+      res.flatMap {
+        case Some(args) =>
+          dapSessionParams(args)
+            .flatMap(debugProvider.asSession(_))
+            .map(Some(_))
+        case None => Future.successful(None)
+      }
+    }
+    private lazy val result: Future[Try[A]] = Future(f())
+    private lazy val retryResult: Future[Option[A]] =
```

Comment:
```suggestion
    private def result: Future[Try[A]] = Future(f())
    private def retryResult: Future[Option[A]] =
```
wouldn't be enough? I am generally wary of lazy val on paralel programming

Reply:
The motivation for separating `ClassSearch` into an entity was so we don't have to recalculate the `result`, when we need `retryResult`. But I can cache `result` explicitly instead of using lazy.

Reply:
Maybe a promise would make sense here?